Add Authentication Indicator Kerberos ticket policy options

abbra opened the pull request:
See discussion in https://github.com/freeipa/freeipa/pull/3358#issuecomment-528796956

Related: https://pagure.io/freeipa/issue/8001

Signed-off-by: Alexander Bokovoy <abokovoy@redhat.com>

abbra commented:
I need to figure out some of object class violations in the tests.

abbra commented:
I updated a test since XMLRPC tests require exact dictionary of returned values.

rcritten commented on daemons/ipa-kdb/ipa_kdb_principals.c:
```diff
@@ -505,6 +507,50 @@ static krb5_error_code ipadb_get_ldap_auth_ind(krb5_context kcontext,
+                                         &result);
+            if (ret == 0) {
+                entry->max_renewable_life = result;
+            }
```
Would it make sense to do a break in here since if we have a match of either the life or age map for an indicator then the others would be guaranteed to not match? I guess it's just over 8 items but still.

abbra replied:
Probably yes, but we have few cases where these are used as flags, so theoretically all of them can be set. It doesn't make sense though...

rcritten commented:
I haven't tracked it down yet but when setting a maxlife I'm getting duplicate multi-value attributes in the ipa cli output but not in the ldapsearch output:
```
ipa krbtpolicy-mod --pkinit-maxlife 380
ipa krbtpolicy-show --all --raw | grep krbAuthIndMaxTicketLi
  krbAuthIndMaxTicketLife;pkinit: 380
  krbAuthIndMaxTicketLife;pkinit: 380
```

abbra commented on ipaserver/plugins/baseldap.py:
```diff
@@ -849,7 +849,11 @@ def _check_limit_object_class(attributes, attrs, allow_only):
+    # LDAP schema does not enforce any of LDAP attribute options
+    # (e.g. attribute;option), thus we should avoid comparing
+    # attribute names with options directly.
+    limitattrs = [x.split(';')[0] for x in attrs]
```
I don't need it, for two reasons:
- attrs is passed in as list(some-dictionary), which makes it a copy already:
```
$ git grep _check_limit_object_class
ipaserver/plugins/baseldap.py:def _check_limit_object_class(attributes, attrs, allow_only):
ipaserver/plugins/baseldap.py:        _check_limit_object_class(self.api.Backend.ldap2.schema.attribute_types(self.obj.limit_object_classes), list(entry_attrs), allow_only=True)
ipaserver/plugins/baseldap.py:        _check_limit_object_class(self.api.Backend.ldap2.schema.attribute_types(self.obj.disallow_object_classes), list(entry_attrs), allow_only=False)
ipaserver/plugins/baseldap.py:        _check_limit_object_class(self.api.Backend.ldap2.schema.attribute_types(self.obj.limit_object_classes), list(entry_attrs), allow_only=True)
ipaserver/plugins/baseldap.py:        _check_limit_object_class(self.api.Backend.ldap2.schema.attribute_types(self.obj.disallow_object_classes), list(entry_attrs), allow_only=False)
```
- `x.split(';')[0]` is already a separate copy

abbra commented:
> I haven't tracked it down yet but when setting a maxlife I'm getting duplicate multi-value attributes in the ipa cli output but not in the ldapsearch output:
> 
> ```
> ipa krbtpolicy-mod --pkinit-maxlife 380
> ipa krbtpolicy-show --all --raw | grep krbAuthIndMaxTicketLi
>   krbAuthIndMaxTicketLife;pkinit: 380
>   krbAuthIndMaxTicketLife;pkinit: 380
> ```
This is, actually, a problem outside of this patchset. I also encountered it and it is CoS plugin behavior where it merges values from the entry and from the template. I guess we need to apply some deeper logic here. I'm getting this result on ldapsearch as well if I specify '*' like our LDAPRetrieve code does:
```
ldapsearch -b uid=foobar,cn=users,cn=accounts,dc=example,dc=test krbMaxRenewableAge krbAuthIndMaxRenewableAge krbMaxTicketLife '*' krbAuthIndMaxTicketLife aci
SASL/GSSAPI authentication started
SASL username: admin@EXAMPLE.TEST
SASL SSF: 256
SASL data security layer installed.
# extended LDIF
#
# LDAPv3
# base <uid=foobar,cn=users,cn=accounts,dc=example,dc=test> with scope subtree
# filter: (objectclass=*)
# requesting: krbMaxRenewableAge krbAuthIndMaxRenewableAge krbMaxTicketLife * krbAuthIndMaxTicketLife aci 
#

# foobar, users, accounts, example.test
dn: uid=foobar,cn=users,cn=accounts,dc=example,dc=test
givenName: Foo
sn: Bar
uid: foobar
cn: Foo Bar
displayName: Foo Bar
initials: FB
gecos: Foo Bar
krbPrincipalName: foobar@EXAMPLE.TEST
objectClass: top
objectClass: person
objectClass: organizationalperson
objectClass: inetorgperson
objectClass: inetuser
objectClass: posixaccount
objectClass: krbprincipalaux
objectClass: krbticketpolicyaux
objectClass: ipaobject
objectClass: ipasshuser
objectClass: ipaSshGroupOfPubKeys
objectClass: mepOriginEntry
loginShell: /bin/sh
homeDirectory: /home/foobar
mail: foobar@example.test
krbCanonicalName: foobar@EXAMPLE.TEST
ipaUniqueID: c9b62c44-d3dc-11e9-b284-fa163e54a9c9
uidNumber: 1821400001
gidNumber: 1821400001
mepManagedEntry: cn=foobar,cn=groups,cn=accounts,dc=example,dc=test
memberOf: cn=ipausers,cn=groups,cn=accounts,dc=example,dc=test
krbAuthIndMaxTicketLife;pkinit: 120000
krbAuthIndMaxTicketLife;otp: 12000
krbMaxTicketLife: 120000
krbAuthIndMaxTicketLife;pkinit: 120000
krbAuthIndMaxTicketLife;otp: 12000

# search result
search: 4
result: 0 Success

# numResponses: 2
# numEntries: 1

```

abbra commented:
While the duplication part seems to be a problem with LDAP server response, we need to re-do actual loading of the policy in the KDB plugin and move application of limits to the `ipa_kdcpolicy_check_as()`. In the `ipadb_parse_ldap_entry` we need to load values and store them in the principal entry struct, then apply them in the `ipa_kdcpolicy_check_as()`

abbra commented:
I fixed actual application of the policy. I had debug logging added in my test environment, when I first create a user and set its Kerberos ticket policy to have a separate setting for 'hardened', it then applies the limits:
```
# ipa krbtpolicy-mod foobar --hardened-maxlife 600
  Hardened max life: 600
# ipa user-mod foobar --user-auth-type={password,hardened}
----------------------
Modified user "foobar"
----------------------
  User login: foobar
  First name: Foo
  Last name: Bar
  Home directory: /home/foobar
  Login shell: /bin/sh
  Principal name: foobar@EXAMPLE.TEST
  Principal alias: foobar@EXAMPLE.TEST
  Email address: foobar@example.test
  UID: 1875200001
  GID: 1875200001
  User authentication types: password, hardened
  Account disabled: False
  Password: True
  Member of groups: ipausers
  Kerberos keys available: True
# ipa config-mod --user-auth-type={password,hardened}
  Maximum username length: 32
  Maximum hostname length: 64
  Home directory base: /home
  Default shell: /bin/sh
  Default users group: ipausers
  Default e-mail domain: example.test
  Search time limit: 2
  Search size limit: 100
  User search fields: uid,givenname,sn,telephonenumber,ou,title
  Group search fields: cn,description
  Enable migration mode: FALSE
  Certificate Subject base: O=EXAMPLE.TEST
  Password Expiration Notification (days): 4
  Password plugin features: AllowNThash, KDC:Disable Last Success
  SELinux user map order: guest_u:s0$xguest_u:s0$user_u:s0$staff_u:s0-s0:c0.c1023$sysadm_u:s0-s0:c0.c1023$unconfined_u:s0-s0:c0.c1023
  Default SELinux user: unconfined_u:s0-s0:c0.c1023
  Default PAC types: MS-PAC, nfs:NONE
  Default user authentication types: password, hardened
  IPA masters: master.example.test
  IPA master capable of PKINIT: master.example.test
  IPA CA servers: master.example.test
  IPA CA renewal master: master.example.test
  IPA DNS servers: master.example.test

# ...
# kinit foobar
Password for foobar@EXAMPLE.TEST: 
# klist -C
Ticket cache: KCM:0:55945
Default principal: foobar@EXAMPLE.TEST

Valid starting       Expires              Service principal
config: fast_avail(krbtgt/EXAMPLE.TEST@EXAMPLE.TEST) = yes
11.09.2019 03.49.54  11.09.2019 03.59.54  krbtgt/EXAMPLE.TEST@EXAMPLE.TEST
config: pa_type(krbtgt/EXAMPLE.TEST@EXAMPLE.TEST) = 151

# cat /var/log/krb5kdc.log
...
Sep 11 03:49:54 master.example.test krb5kdc[28007](info): IPA kdcpolicy: checking AS-REQ.
Sep 11 03:49:54 master.example.test krb5kdc[28007](info): IPA kdcpolicy: checking indicator hardened.
Sep 11 03:49:54 master.example.test krb5kdc[28007](info): IPA kdcpolicy: policy limites are max life 600 and max renewal age 0.
Sep 11 03:49:54 master.example.test krb5kdc[28007](info): AS_REQ (8 etypes {aes256-cts-hmac-sha1-96(18), aes128-cts-hmac-sha1-96(17), aes256-cts-hmac-sha384-192(20), aes128-cts-hmac-sha256-128(19), DEPRECATED:des3-cbc-sha1(16), DEPRECATED:arcfour-hmac(23), camellia128-cts-cmac(25), camellia256-cts-cmac(26)}) 2620:52:0:88:f816:3eff:fe6d:ac78: ISSUE: authtime 1568188194, etypes {rep=aes256-cts-hmac-sha1-96(18), tkt=aes256-cts-hmac-sha1-96(18), ses=aes256-cts-hmac-sha1-96(18)}, foobar@EXAMPLE.TEST for krbtgt/EXAMPLE.TEST@EXAMPLE.TEST
...
```

You can see with `klist -C` that the pre-authentication type is 151 (SPAKE, aka 'hardened').

frozencemetery commented on daemons/ipa-kdb/ipa_kdb_principals.c:
```diff
@@ -505,6 +509,66 @@ static krb5_error_code ipadb_get_ldap_auth_ind(krb5_context kcontext,
+    const struct {
+        char *attribute;
+        enum ipadb_user_auth flag;
+	enum ipadb_user_auth_idx idx;
```
indentation

frozencemetery commented on daemons/ipa-kdb/ipa_kdb_principals.c:
```diff
@@ -505,6 +509,66 @@ static krb5_error_code ipadb_get_ldap_auth_ind(krb5_context kcontext,
+        return;
+    }
+
+    for(size_t i = 0; life_authind_map[i].attribute != NULL; i++) {
```
space after `for`

abbra commented:
Fixed formatting problems pointed by @frozencemetery and also removed subattributes from the ACIs, as pointed by @simo5 on IRC.

rcritten commented:
I can't reproduce the limit being applied. I followed your reproduction steps for foobar but what I get is:

```
# ipa krbtpolicy-show foobar
  Max life: 86400
  Max renew: 604800
  Hardened max life: 600
# kinit foobar
# klist -C
Ticket cache: KEYRING:persistent:0:krb_ccache_smNBRcI
Default principal: foobar@EXAMPLE.TEST

Valid starting       Expires              Service principal
config: pa_type(krbtgt/EXAMPLE.TEST@EXAMPLE.TEST) = 151
config: fast_avail(krbtgt/EXAMPLE.TEST@EXAMPLE.TEST) = yes
09/11/2019 16:08:17  09/12/2019 16:08:15  krbtgt/EXAMPLE.TEST@EXAMPLE.TEST
```

abbra commented:
Did you enable hardened for both the user and globally in user  authentication types?

frozencemetery commented:
So this seems to have stalled.  What's the next step?  Does anyone have ideas what might be amiss that Rob found?

abbra commented:
Rebased on top of the current git master.

abbra commented:
@rcritten Just a thought -- may be you could try with a different ccache type? In my test it worked with KCM -- your output shows keyring in use.

rcritten commented:
I re-installed, re-enabled KCM and got a 10-minute ticket.

As a test I disabled KCM again, kdestroy -A, kinit and again got a 10 minute ticket.

I literally cut-and-pasted the steps I used before so I'm not sure why it failed then and not now but the code appears to be working properly.

abbra commented:
Ok, since the test procedure is working, one thing left is to write it down as an integration test.

rcritten commented:
@frozencemetery is the test that something you think you could add or do you need one of the IPA developers to help?

frozencemetery commented:
I probably could do this, but note that it would be a separate PR (since abbra was kind enough to author this one).  It also won't happen for a bit since I need to work on other things.

So I'd prefer someone else (abbra?) handle it, but can do it myself if it comes to that.

rcritten commented:
I can work up a test and push to @abbra branch as an additional commit.

abbra commented:
@rcritten awesome, thank you very much. I checked the results of runs and they are fine. Ticket lifetime was properly reduced in all tests where it was expected.
Please convert the temp commit into a proper test addition.

rcritten commented:
@abbra it is already configured to run in the nightly builds. Did you want to run this with each PR? I thought that would be overkill.

rcritten commented:
Added missing test config to nightly_latest_testing.yaml and dropped the temp commit.

flo-renaud commented:
- `fedora-latest/dnssec` failure is a known issue ([7985](https://pagure.io/freeipa/issue/7985)), you can relaunch the test once the conflicts on nightly yaml files are resolved.

flo-renaud commented:
The tests LGTM, and the server code was already reviewed by @rcritten hence ACK.

rcritten commented:
master:

* 9db6f65a8536781f6bca20ff8031de18c0f79397 Allow presence of LDAP attribute options
* c5f32165d6105a48d9de85a8d29925b58beb9f91 Add Authentication Indicator Kerberos ticket policy options
* c02cc93c146bca429f69bf5536a3f6c15b02876e Add integration test for Kerberos ticket policy
